src/psbt/mod.rs

```diff
@@ -366,32 +371,408 @@ fn sanity_check(psbt: &Psbt) -> Result<(), Error> {
+            }
+            UxtoUpdateError::DerivationError(e) => write!(f, "Key Derivation error {}", e),
+            UxtoUpdateError::MissingInputUxto => write!(f, "Missing input utxo in pbst"),
+            UxtoUpdateError::RangeMissingWildCardDescriptor => {
+                write!(
+                    f,
+                    "Range missing must be supplied for a wild-card descriptor"
+                )
+            }
+            UxtoUpdateError::IncorrectDescriptor => {
+                write!(
+                    f,
+                    "Cannot derive the output script pubkey using the given descriptor and range"
+                )
+            }
         }
+    }
+}
+
+impl error::Error for UxtoUpdateError {}
+
+/// Return error type for [`PbstOps::sighash_msg`]
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Copy)]
+pub enum SigHashError {
+    /// Index out of bounds
+    IndexOutOfBounds(usize, usize),
+    /// Missing input utxo
+    MissingInputUxto,
+    /// Missing Prevouts
+    MissingSpendUtxos,
+    /// Invalid Sighash type
+    InvalidSigHashType,
+    /// Sighash computation error
+    /// Only happens when single does not have corresponding output as psbts
+    /// already have information to compute the sighash
+    SigHashComputationError(bitcoin::util::sighash::Error),
+    /// Missing Witness script
+    MissingWitnessScript,
+    /// Missing Redeem script,
+    MissingRedeemScript,
+}
 
-        if let Some(witness) = input.final_script_witness.as_ref() {
-            ret.input[n].witness = witness.clone();
+impl fmt::Display for SigHashError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            SigHashError::IndexOutOfBounds(ind, len) => {
+                write!(f, "index {}, psbt input len: {}", ind, len)
+            }
+            SigHashError::MissingInputUxto => write!(f, "Missing input utxo in pbst"),
+            SigHashError::MissingSpendUtxos => write!(f, "Missing Psbt spend utxos"),
+            SigHashError::InvalidSigHashType => write!(f, "Invalid Sighash type"),
+            SigHashError::SigHashComputationError(e) => {
+                write!(f, "Sighash computation error : {}", e)
+            }
+            SigHashError::MissingWitnessScript => write!(f, "Missing Witness Script"),
+            SigHashError::MissingRedeemScript => write!(f, "Missing Redeem Script"),
         }
-        if let Some(script_sig) = input.final_script_sig.as_ref() {
-            ret.input[n].script_sig = script_sig.clone();
+    }
+}
+
+impl From<bitcoin::util::sighash::Error> for SigHashError {
+    fn from(e: bitcoin::util::sighash::Error) -> Self {
+        SigHashError::SigHashComputationError(e)
+    }
+}
+
+impl error::Error for SigHashError {}
+
+/// Sighash message(signing data) for a given psbt transaction input.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Copy)]
+pub enum PsbtSigHashMsg {
```

Comment:
In 27679de0b4ba07da7e007b1bf046834ebf0abae2:

I think these variants should also hold the sighash type. Thoughts?

Reply:
Why do you think this might be useful? The sighash type is directly available in psbt APIs by https://github.com/rust-bitcoin/rust-bitcoin/pull/847/ . The signer only needs to know the message in order to sign with keys

Reply:
Ah, good point. I forgot about the other accessor function.